File: LeetCode/Graph/DFS/JS/Leetcode919-complete-binary-tree-inserter.js
var CBTInserter = function(root) {
    this.rec = [new TreeNode()];
    this.root = root;
    let q = [root];
    while(q.length) {
        let t = q.shift();
        this.rec.push(t);
        if(t.left) q.push(t.left);
        if(t.right) q.push(t.right);
    }

};

/** 
 * @param {number} val
 * @return {number}
 */
CBTInserter.prototype.insert = function(val) {
    let node = new TreeNode(val);
    this.rec.push(node);
    let k = this.rec.length - 1;
    let p = parseInt(k / 2);
    if(p * 2 == k) {
        this.rec[p].left = node;
    }else{
        this.rec[p].right = node;
    }
    return this.rec[p].val;
};

/**
 * @return {TreeNode}
 */
CBTInserter.prototype.get_root = function() {
    return this.root;
};

File: LeetCode/Graph/DFS/Golang/Leetcode919-complete-binary-tree-inserter.go
package sbgo

type CBTInserter struct
{
  rec []*TreeNode
  root *TreeNode
}   


func Constructor(root *TreeNode) CBTInserter {
  rec := []*TreeNode{nil,root}
  q := []*TreeNode{root}
  for len(q) != 0 {
    t := q[0]
    q = q[1:]
    rec = append(rec,t)
    if t.Left != nil {
      q = append(q,t.Left)
    }
    if t.Right != nil {
      q = append(q,t.Right)
    }
  }
  return CBTInserter{rec,root}
}


func (this *CBTInserter) Insert(val int) int {
  node := &TreeNode{val,nil,nil}
  this.rec = append(this.rec,node)
  k := len(this.rec)
  p := k / 2;
  if p * 2 == k {
    this.rec[p].Left = node
  }else{
    this.rec[p].Right = node
  }
  return this.rec[p].Val
}


func (this *CBTInserter) Get_root() *TreeNode {
  return this.root
}

